packages/doenetml-worker-rust/lib-doenetml-core/src/core/state/types/math_expr.rs

```diff
@@ -0,0 +1,429 @@
+use std::collections::HashMap;
+use strum_macros::Display;
+
+use crate::math_via_wasm::{
+    eval_js, math_to_latex, normalize_math, parse_latex_into_math, parse_text_into_math,
+    substitute_into_math,
+};
+
+const BLANK_MATH_OBJECT: &str = "{\"objectType\":\"math-expression\",\"tree\":\"\u{ff3f}\"}";
+/// A mathematical expression represented as a serialized `math-expressions` object.
+#[derive(Debug, Clone, PartialEq, serde::Deserialize)]
+pub struct MathExpr {
+    pub math_object: String,
+}
+
+impl Default for MathExpr {
+    fn default() -> Self {
+        MathExpr {
+            math_object: BLANK_MATH_OBJECT.to_string(),
+        }
+    }
+}
+
+impl std::fmt::Display for MathExpr {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.to_latex(ToLatexParams::default()))
+    }
+}
+
+impl serde::Serialize for MathExpr {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_str(&self.math_object)
+    }
+}
+
+impl MathExpr {
+    /// Create a `Math` by parsing the string `text` using the `math-expressions` text parser.
+    ///
+    /// Parameters:
+    /// - `text`: the string to be parsed
+    /// - `split_symbols`: if true, split multi-character symbols that do not contain an number
+    ///   into the product of the characters.
+    /// - `function_symbols`: a list of the symbols that will be treated as a function,
+    ///   i.e., one of these symbols followed by arguments in parentheses
+    ///   will be interpreted as apply that function to the arguments (rather than multiplication)
+    ///
+    /// Examples:
+    /// - `MathExpr::from_text("xy", true, &["f"])` is `x` times `y`.
```

Comment:
All your examples pass `&["f"]`, which makes it look mandatory.

Reply:
Right, because if you pass a `&[]`, it gets mad at you for not being able to infer the type. So, I gave up and just passed in `&["f"]` instead. It seemed that adding type annotations to `&[]` made it look too complicated. Not sure the best approach here.

Reply:
Maybe make the last argument an option, then? Or perhaps the builder patter would be better here. You could add a `finalize()` option that would initiate the parsing and return the actual object. It would look like

```
MathExpr::from_text("xy").split_symbols().finalize()`
```

Reply:
Making the last argument an option would be the simplest approach.

I thought about the builder pattern, but then it wasn't clear how to minimize the number of calls Javascript. You could imagine wanting to do something like
```
MathExpr::from_text("xy(1+xy)").split_symbols().simplify().expand()
```
and have that communicate once to Javascript and then come back. But that seemed too complicated for now so I stuck with the separate functions, where each function is a call to Javascript. Maybe your `.finalize()` idea takes care of that, but it'd be nice if that weren't required.

Reply:
Ugh, making `function_symbols` an `Option<&[FnSymbol]>` didn't help, as passing it `None` for `function_symbols` still makes it mad because it cannot infer the type of the parameter `FnSymbol`.